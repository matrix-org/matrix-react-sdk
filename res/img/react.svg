<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
