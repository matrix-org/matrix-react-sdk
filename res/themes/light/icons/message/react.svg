<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>