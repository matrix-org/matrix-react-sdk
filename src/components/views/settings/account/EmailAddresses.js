/*
Copyright 2019 New Vector Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import PropTypes from 'prop-types';
import {_t} from "../../../../languageHandler";
import {MatrixClientPeg} from "../../../../MatrixClientPeg";
import Field from "../../elements/Field";
import AccessibleButton from "../../elements/AccessibleButton";
import * as Email from "../../../../email";
import AddThreepid from "../../../../AddThreepid";
import * as sdk from '../../../../index';
import Modal from '../../../../Modal';

/*
TODO: Improve the UX for everything in here.
It's very much placeholder, but it gets the job done. The old way of handling
email addresses in user settings was to use dialogs to communicate state, however
due to our dialog system overriding dialogs (causing unmounts) this creates problems
for a sane UX. For instance, the user could easily end up entering an email address
and receive a dialog to verify the address, which then causes the component here
to forget what it was doing and ultimately fail. Dialogs are still used in some
places to communicate errors - these should be replaced with inline validation when
that is available.
 */

export class ExistingEmailAddress extends React.Component {
    static propTypes = {
        email: PropTypes.object.isRequired,
        onRemoved: PropTypes.func.isRequired,
    };

    constructor() {
        super();

        this.state = {
            verifyRemove: false,
        };
    }

    _onRemove = (e) => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({verifyRemove: true});
    };

    _onDontRemove = (e) => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({verifyRemove: false});
    };

    _onActuallyRemove = (e) => {
        e.stopPropagation();
        e.preventDefault();

        MatrixClientPeg.get().deleteThreePid(this.props.email.medium, this.props.email.address).then(() => {
            return this.props.onRemoved(this.props.email);
        }).catch((err) => {
            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
            console.error("Unable to remove contact information: " + err);
            Modal.createTrackedDialog('Remove 3pid failed', '', ErrorDialog, {
                title: _t("Unable to remove contact information"),
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        });
    };

    render() {
        if (this.state.verifyRemove) {
            return (
                <div className="mx_ExistingEmailAddress">
                    <span className="mx_ExistingEmailAddress_promptText">
                        {_t("Remove %(email)s?", {email: this.props.email.address} )}
                    </span>
                    <AccessibleButton onClick={this._onActuallyRemove} kind="danger_sm"
                                      className="mx_ExistingEmailAddress_confirmBtn">
                        {_t("Remove")}
                    </AccessibleButton>
                    <AccessibleButton onClick={this._onDontRemove} kind="link_sm"
                                      className="mx_ExistingEmailAddress_confirmBtn">
                        {_t("Cancel")}
                    </AccessibleButton>
                </div>
            );
        }

        return (
            <div className="mx_ExistingEmailAddress">
                <span className="mx_ExistingEmailAddress_email">{this.props.email.address}</span>
                <AccessibleButton onClick={this._onRemove} kind="danger_sm">
                    {_t("Remove")}
                </AccessibleButton>
            </div>
        );
    }
}

export default class EmailAddresses extends React.Component {
    static propTypes = {
        emails: PropTypes.array.isRequired,
        onEmailsChange: PropTypes.func.isRequired,
    }

    constructor(props) {
        super(props);

        this.state = {
            verifying: false,
            addTask: null,
            continueDisabled: false,
            newEmailAddress: "",
        };
    }

    _onRemoved = (address) => {
        const emails = this.props.emails.filter((e) => e !== address);
        this.props.onEmailsChange(emails);
    };

    _onChangeNewEmailAddress = (e) => {
        this.setState({
            newEmailAddress: e.target.value,
        });
    };

    _onAddClick = (e) => {
        e.stopPropagation();
        e.preventDefault();

        if (!this.state.newEmailAddress) return;

        const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
        const email = this.state.newEmailAddress;

        // TODO: Inline field validation
        if (!Email.looksValid(email)) {
            Modal.createTrackedDialog('Invalid email address', '', ErrorDialog, {
                title: _t("Invalid Email Address"),
                description: _t("This doesn't appear to be a valid email address"),
            });
            return;
        }

        const task = new AddThreepid();
        this.setState({verifying: true, continueDisabled: true, addTask: task});

        task.addEmailAddress(email).then(() => {
            this.setState({continueDisabled: false});
        }).catch((err) => {
            console.error("Unable to add email address " + email + " " + err);
            this.setState({verifying: false, continueDisabled: false, addTask: null});
            Modal.createTrackedDialog('Unable to add email address', '', ErrorDialog, {
                title: _t("Unable to add email address"),
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        });
    };

    _onContinueClick = (e) => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({continueDisabled: true});
        this.state.addTask.checkEmailLinkClicked().then(() => {
            const email = this.state.newEmailAddress;
            this.setState({
                addTask: null,
                continueDisabled: false,
                verifying: false,
                newEmailAddress: "",
            });
            const emails = [
                ...this.props.emails,
                { address: email, medium: "email" },
            ];
            this.props.onEmailsChange(emails);
        }).catch((err) => {
            this.setState({continueDisabled: false});
            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
            if (err.errcode === 'M_THREEPID_AUTH_FAILED') {
                Modal.createTrackedDialog("Email hasn't been verified yet", "", ErrorDialog, {
                    title: _t("Your email address hasn't been verified yet"),
                    description: _t("Click the link in the email you received to verify " +
                        "and then click continue again."),
                });
            } else {
                console.error("Unable to verify email address: ", err);
                Modal.createTrackedDialog('Unable to verify email address', '', ErrorDialog, {
                    title: _t("Unable to verify email address."),
                    description: ((err && err.message) ? err.message : _t("Operation failed")),
                });
            }
        });
    };

    render() {
        const existingEmailElements = this.props.emails.map((e) => {
            return <ExistingEmailAddress email={e} onRemoved={this._onRemoved} key={e.address} />;
        });

        let addButton = (
            <AccessibleButton onClick={this._onAddClick} kind="primary">
                {_t("Add")}
            </AccessibleButton>
        );
        if (this.state.verifying) {
            addButton = (
              <div>
                  <div>{_t("We've sent you an email to verify your address. Please follow the instructions there and then click the button below.")}</div>
                  <AccessibleButton onClick={this._onContinueClick} kind="primary"
                                    disabled={this.state.continueDisabled}>
                      {_t("Continue")}
                  </AccessibleButton>
              </div>
            );
        }

        return (
            <div className="mx_EmailAddresses">
                {existingEmailElements}
                <form onSubmit={this._onAddClick} autoComplete="off"
                      noValidate={true} className="mx_EmailAddresses_new">
                    <Field id="mx_EmailAddressses_newEmailAddress"
                        type="text"
                        label={_t("Email Address")}
                        autoComplete="off"
                        disabled={this.state.verifying}
                        value={this.state.newEmailAddress}
                        onChange={this._onChangeNewEmailAddress}
                    />
                    {addButton}
                </form>
            </div>
        );
    }
}
