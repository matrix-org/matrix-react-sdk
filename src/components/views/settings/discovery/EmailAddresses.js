/*
Copyright 2019 New Vector Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import PropTypes from 'prop-types';

import { _t } from "../../../../languageHandler";
import {MatrixClientPeg} from "../../../../MatrixClientPeg";
import * as sdk from '../../../../index';
import Modal from '../../../../Modal';
import AddThreepid from '../../../../AddThreepid';
import {replaceableComponent} from "../../../../utils/replaceableComponent";

/*
TODO: Improve the UX for everything in here.
It's very much placeholder, but it gets the job done. The old way of handling
email addresses in user settings was to use dialogs to communicate state, however
due to our dialog system overriding dialogs (causing unmounts) this creates problems
for a sane UX. For instance, the user could easily end up entering an email address
and receive a dialog to verify the address, which then causes the component here
to forget what it was doing and ultimately fail. Dialogs are still used in some
places to communicate errors - these should be replaced with inline validation when
that is available.
*/

/*
TODO: Reduce all the copying between account vs. discovery components.
*/

export class EmailAddress extends React.Component {
    static propTypes = {
        email: PropTypes.object.isRequired,
    };

    constructor(props) {
        super(props);

        const { bound } = props.email;

        this.state = {
            verifying: false,
            addTask: null,
            continueDisabled: false,
            bound,
        };
    }

    // TODO: [REACT-WARNING] Replace with appropriate lifecycle event
    UNSAFE_componentWillReceiveProps(nextProps) { // eslint-disable-line camelcase
        const { bound } = nextProps.email;
        this.setState({ bound });
    }

    async changeBinding({ bind, label, errorTitle }) {
        if (!await MatrixClientPeg.get().doesServerSupportSeparateAddAndBind()) {
            return this.changeBindingTangledAddBind({ bind, label, errorTitle });
        }

        const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
        const { medium, address } = this.props.email;

        try {
            if (bind) {
                const task = new AddThreepid();
                this.setState({
                    verifying: true,
                    continueDisabled: true,
                    addTask: task,
                });
                await task.bindEmailAddress(address);
                this.setState({
                    continueDisabled: false,
                });
            } else {
                await MatrixClientPeg.get().unbindThreePid(medium, address);
            }
            this.setState({ bound: bind });
        } catch (err) {
            console.error(`Unable to ${label} email address ${address} ${err}`);
            this.setState({
                verifying: false,
                continueDisabled: false,
                addTask: null,
            });
            Modal.createTrackedDialog(`Unable to ${label} email address`, '', ErrorDialog, {
                title: errorTitle,
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        }
    }

    async changeBindingTangledAddBind({ bind, label, errorTitle }) {
        const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
        const { medium, address } = this.props.email;

        const task = new AddThreepid();
        this.setState({
            verifying: true,
            continueDisabled: true,
            addTask: task,
        });

        try {
            await MatrixClientPeg.get().deleteThreePid(medium, address);
            if (bind) {
                await task.bindEmailAddress(address);
            } else {
                await task.addEmailAddress(address);
            }
            this.setState({
                continueDisabled: false,
                bound: bind,
            });
        } catch (err) {
            console.error(`Unable to ${label} email address ${address} ${err}`);
            this.setState({
                verifying: false,
                continueDisabled: false,
                addTask: null,
            });
            Modal.createTrackedDialog(`Unable to ${label} email address`, '', ErrorDialog, {
                title: errorTitle,
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        }
    }

    onRevokeClick = (e) => {
        e.stopPropagation();
        e.preventDefault();
        this.changeBinding({
            bind: false,
            label: "revoke",
            errorTitle: _t("Unable to revoke sharing for email address"),
        });
    }

    onShareClick = (e) => {
        e.stopPropagation();
        e.preventDefault();
        this.changeBinding({
            bind: true,
            label: "share",
            errorTitle: _t("Unable to share email address"),
        });
    }

    onContinueClick = async (e) => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({ continueDisabled: true });
        try {
            await this.state.addTask.checkEmailLinkClicked();
            this.setState({
                addTask: null,
                continueDisabled: false,
                verifying: false,
            });
        } catch (err) {
            this.setState({ continueDisabled: false });
            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
            if (err.errcode === 'M_THREEPID_AUTH_FAILED') {
                Modal.createTrackedDialog("E-mail hasn't been verified yet", "", ErrorDialog, {
                    title: _t("Your email address hasn't been verified yet"),
                    description: _t("Click the link in the email you received to verify " +
                        "and then click continue again."),
                });
            } else {
                console.error("Unable to verify email address: " + err);
                Modal.createTrackedDialog('Unable to verify email address', '', ErrorDialog, {
                    title: _t("Unable to verify email address."),
                    description: ((err && err.message) ? err.message : _t("Operation failed")),
                });
            }
        }
    }

    render() {
        const AccessibleButton = sdk.getComponent('elements.AccessibleButton');
        const { address } = this.props.email;
        const { verifying, bound } = this.state;

        let status;
        if (verifying) {
            status = <span>
                {_t("Verify the link in your inbox")}
                <AccessibleButton
                    className="mx_ExistingEmailAddress_confirmBtn"
                    kind="primary_sm"
                    onClick={this.onContinueClick}
                    disabled={this.state.continueDisabled}
                >
                    {_t("Complete")}
                </AccessibleButton>
            </span>;
        } else if (bound) {
            status = <AccessibleButton
                className="mx_ExistingEmailAddress_confirmBtn"
                kind="danger_sm"
                onClick={this.onRevokeClick}
            >
                {_t("Revoke")}
            </AccessibleButton>;
        } else {
            status = <AccessibleButton
                className="mx_ExistingEmailAddress_confirmBtn"
                kind="primary_sm"
                onClick={this.onShareClick}
            >
                {_t("Share")}
            </AccessibleButton>;
        }

        return (
            <div className="mx_ExistingEmailAddress">
                <span className="mx_ExistingEmailAddress_email">{address}</span>
                {status}
            </div>
        );
    }
}

@replaceableComponent("views.settings.discovery.EmailAddresses")
export default class EmailAddresses extends React.Component {
    static propTypes = {
        emails: PropTypes.array.isRequired,
    }

    render() {
        let content;
        if (this.props.emails.length > 0) {
            content = this.props.emails.map((e) => {
                return <EmailAddress email={e} key={e.address} />;
            });
        } else {
            content = <span className="mx_SettingsTab_subsectionText">
                {_t("Discovery options will appear once you have added an email above.")}
            </span>;
        }

        return (
            <div className="mx_EmailAddresses">
                {content}
            </div>
        );
    }
}
