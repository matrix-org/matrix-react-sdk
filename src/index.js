/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
