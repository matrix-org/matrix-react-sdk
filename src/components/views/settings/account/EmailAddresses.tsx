/*
Copyright 2019 New Vector Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { IThreepid, ThreepidMedium } from "matrix-js-sdk/src/@types/threepids";
import { logger } from "matrix-js-sdk/src/logger";

import { _t } from "../../../../languageHandler";
import { MatrixClientPeg } from "../../../../MatrixClientPeg";
import Field from "../../elements/Field";
import AccessibleButton from "../../elements/AccessibleButton";
import * as Email from "../../../../email";
import AddThreepid from "../../../../AddThreepid";
import Modal from '../../../../Modal';
import ErrorDialog from "../../dialogs/ErrorDialog";

/*
TODO: Improve the UX for everything in here.
It's very much placeholder, but it gets the job done. The old way of handling
email addresses in user settings was to use dialogs to communicate state, however
due to our dialog system overriding dialogs (causing unmounts) this creates problems
for a sane UX. For instance, the user could easily end up entering an email address
and receive a dialog to verify the address, which then causes the component here
to forget what it was doing and ultimately fail. Dialogs are still used in some
places to communicate errors - these should be replaced with inline validation when
that is available.
 */

interface IExistingEmailAddressProps {
    email: IThreepid;
    onRemoved: (emails: IThreepid) => void;
}

interface IExistingEmailAddressState {
    verifyRemove: boolean;
}

export class ExistingEmailAddress extends React.Component<IExistingEmailAddressProps, IExistingEmailAddressState> {
    constructor(props: IExistingEmailAddressProps) {
        super(props);

        this.state = {
            verifyRemove: false,
        };
    }

    private onRemove = (e: React.MouseEvent): void => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({ verifyRemove: true });
    };

    private onDontRemove = (e: React.MouseEvent): void => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({ verifyRemove: false });
    };

    private onActuallyRemove = (e: React.MouseEvent): void => {
        e.stopPropagation();
        e.preventDefault();

        MatrixClientPeg.get().deleteThreePid(this.props.email.medium, this.props.email.address).then(() => {
            return this.props.onRemoved(this.props.email);
        }).catch((err) => {
            logger.error("Unable to remove contact information: " + err);
            Modal.createDialog(ErrorDialog, {
                title: _t("Unable to remove contact information"),
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        });
    };

    public render(): JSX.Element {
        if (this.state.verifyRemove) {
            return (
                <div className="mx_ExistingEmailAddress">
                    <span className="mx_ExistingEmailAddress_promptText">
                        { _t("Remove %(email)s?", { email: this.props.email.address }) }
                    </span>
                    <AccessibleButton
                        onClick={this.onActuallyRemove}
                        kind="danger_sm"
                        className="mx_ExistingEmailAddress_confirmBtn"
                    >
                        { _t("Remove") }
                    </AccessibleButton>
                    <AccessibleButton
                        onClick={this.onDontRemove}
                        kind="link_sm"
                        className="mx_ExistingEmailAddress_confirmBtn"
                    >
                        { _t("Cancel") }
                    </AccessibleButton>
                </div>
            );
        }

        return (
            <div className="mx_ExistingEmailAddress">
                <span className="mx_ExistingEmailAddress_email">{ this.props.email.address }</span>
                <AccessibleButton onClick={this.onRemove} kind="danger_sm">
                    { _t("Remove") }
                </AccessibleButton>
            </div>
        );
    }
}

interface IProps {
    emails: IThreepid[];
    onEmailsChange: (emails: Partial<IThreepid>[]) => void;
}

interface IState {
    verifying: boolean;
    addTask: any; // FIXME: When AddThreepid is TSfied
    continueDisabled: boolean;
    newEmailAddress: string;
}

export default class EmailAddresses extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.state = {
            verifying: false,
            addTask: null,
            continueDisabled: false,
            newEmailAddress: "",
        };
    }

    private onRemoved = (address): void => {
        const emails = this.props.emails.filter((e) => e !== address);
        this.props.onEmailsChange(emails);
    };

    private onChangeNewEmailAddress = (e: React.ChangeEvent<HTMLInputElement>): void => {
        this.setState({
            newEmailAddress: e.target.value,
        });
    };

    private onAddClick = (e: React.FormEvent): void => {
        e.stopPropagation();
        e.preventDefault();

        if (!this.state.newEmailAddress) return;

        const email = this.state.newEmailAddress;

        // TODO: Inline field validation
        if (!Email.looksValid(email)) {
            Modal.createDialog(ErrorDialog, {
                title: _t("Invalid Email Address"),
                description: _t("This doesn't appear to be a valid email address"),
            });
            return;
        }

        const task = new AddThreepid();
        this.setState({ verifying: true, continueDisabled: true, addTask: task });

        task.addEmailAddress(email).then(() => {
            this.setState({ continueDisabled: false });
        }).catch((err) => {
            logger.error("Unable to add email address " + email + " " + err);
            this.setState({ verifying: false, continueDisabled: false, addTask: null });
            Modal.createDialog(ErrorDialog, {
                title: _t("Unable to add email address"),
                description: ((err && err.message) ? err.message : _t("Operation failed")),
            });
        });
    };

    private onContinueClick = (e: React.MouseEvent): void => {
        e.stopPropagation();
        e.preventDefault();

        this.setState({ continueDisabled: true });
        this.state.addTask.checkEmailLinkClicked().then(([finished]) => {
            let newEmailAddress = this.state.newEmailAddress;
            if (finished) {
                const email = this.state.newEmailAddress;
                const emails = [
                    ...this.props.emails,
                    { address: email, medium: ThreepidMedium.Email },
                ];
                this.props.onEmailsChange(emails);
                newEmailAddress = "";
            }
            this.setState({
                addTask: null,
                continueDisabled: false,
                verifying: false,
                newEmailAddress,
            });
        }).catch((err) => {
            this.setState({ continueDisabled: false });
            if (err.errcode === 'M_THREEPID_AUTH_FAILED') {
                Modal.createDialog(ErrorDialog, {
                    title: _t("Your email address hasn't been verified yet"),
                    description: _t("Click the link in the email you received to verify " +
                        "and then click continue again."),
                });
            } else {
                logger.error("Unable to verify email address: ", err);
                Modal.createDialog(ErrorDialog, {
                    title: _t("Unable to verify email address."),
                    description: ((err && err.message) ? err.message : _t("Operation failed")),
                });
            }
        });
    };

    public render(): JSX.Element {
        const existingEmailElements = this.props.emails.map((e) => {
            return <ExistingEmailAddress email={e} onRemoved={this.onRemoved} key={e.address} />;
        });

        let addButton = (
            <AccessibleButton onClick={this.onAddClick} kind="primary">
                { _t("Add") }
            </AccessibleButton>
        );
        if (this.state.verifying) {
            addButton = (
                <div>
                    <div>{ _t("We've sent you an email to verify your address. Please follow the instructions there and then click the button below.") }</div>
                    <AccessibleButton
                        onClick={this.onContinueClick}
                        kind="primary"
                        disabled={this.state.continueDisabled}
                    >
                        { _t("Continue") }
                    </AccessibleButton>
                </div>
            );
        }

        return (
            <div className="mx_EmailAddresses">
                { existingEmailElements }
                <form
                    onSubmit={this.onAddClick}
                    autoComplete="off"
                    noValidate={true}
                    className="mx_EmailAddresses_new"
                >
                    <Field
                        type="text"
                        label={_t("Email Address")}
                        autoComplete="off"
                        disabled={this.state.verifying}
                        value={this.state.newEmailAddress}
                        onChange={this.onChangeNewEmailAddress}
                    />
                    { addButton }
                </form>
            </div>
        );
    }
}
