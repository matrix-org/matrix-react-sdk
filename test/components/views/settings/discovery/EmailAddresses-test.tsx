/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { IThreepid, ThreepidMedium, IRequestTokenResponse, MatrixError } from "matrix-js-sdk/src/matrix";

import { TranslationKey, UserFriendlyError } from "../../../../../src/languageHandler";
import EmailAddresses, { EmailAddress } from "../../../../../src/components/views/settings/discovery/EmailAddresses";
import { clearAllModals, getMockClientWithEventEmitter } from "../../../../test-utils";

const mockGetAccessToken = jest.fn().mockResolvedValue("getAccessToken");
jest.mock("../../../../../src/IdentityAuthClient", () =>
    jest.fn().mockImplementation(() => ({
        getAccessToken: mockGetAccessToken,
    })),
);

const emailThreepidFixture: IThreepid = {
    medium: ThreepidMedium.Email,
    address: "foo@bar.com",
    validated_at: 12345,
    added_at: 12342,
    bound: false,
};

describe("<EmailAddress/>", () => {
    const mockClient = getMockClientWithEventEmitter({
        getIdentityServerUrl: jest.fn().mockReturnValue("https://fake-identity-server"),
        generateClientSecret: jest.fn(),
        requestEmailToken: jest.fn(),
        bindThreePid: jest.fn(),
    });

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(async () => {
        jest.useRealTimers();
        await clearAllModals();
    });

    it("should track props.email.bound changes", async () => {
        const { rerender } = render(<EmailAddress email={emailThreepidFixture} />);
        await screen.findByText("Share");

        rerender(
            <EmailAddress
                email={{
                    ...emailThreepidFixture,
                    bound: true,
                }}
            />,
        );
        await screen.findByText("Revoke");
    });

    describe("Email verification share phase", () => {
        it("shows translated error message", async () => {
            render(<EmailAddress email={emailThreepidFixture} />);
            mockClient.requestEmailToken.mockRejectedValue(
                new MatrixError(
                    { errcode: "M_THREEPID_IN_USE", error: "Some fake MatrixError occured" },
                    400,
                    "https://fake-url/",
                ),
            );
            fireEvent.click(screen.getByText("Share"));

            // Expect error dialog/modal to be shown. We have to wait for the UI to transition.
            expect(await screen.findByText("This email address is already in use")).toBeInTheDocument();
        });
    });

    describe("Email verification complete phase", () => {
        beforeEach(async () => {
            // Start these tests out at the "Complete" phase
            render(<EmailAddress email={emailThreepidFixture} />);
            mockClient.requestEmailToken.mockResolvedValue({ sid: "123-fake-sid" } satisfies IRequestTokenResponse);
            fireEvent.click(screen.getByText("Share"));
            // Then wait for the completion screen to come up
            await screen.findByText("Complete");
        });

        it("Shows error dialog when share completion fails (email not verified yet)", async () => {
            mockClient.bindThreePid.mockRejectedValue(
                new MatrixError(
                    { errcode: "M_THREEPID_AUTH_FAILED", error: "Some fake MatrixError occured" },
                    403,
                    "https://fake-url/",
                ),
            );
            fireEvent.click(screen.getByText("Complete"));

            // Expect error dialog/modal to be shown. We have to wait for the UI to transition.
            // Check the title
            expect(await screen.findByText("Your email address hasn't been verified yet")).toBeInTheDocument();
            // Check the description
            expect(
                await screen.findByText(
                    "Click the link in the email you received to verify and then click continue again.",
                ),
            ).toBeInTheDocument();
        });

        it("Shows error dialog when share completion fails (UserFriendlyError)", async () => {
            const fakeErrorText = "Fake UserFriendlyError error in test" as TranslationKey;
            mockClient.bindThreePid.mockRejectedValue(new UserFriendlyError(fakeErrorText));
            fireEvent.click(screen.getByText("Complete"));

            // Expect error dialog/modal to be shown. We have to wait for the UI to transition.
            // Check the title
            expect(await screen.findByText("Unable to verify email address.")).toBeInTheDocument();
            // Check the description
            expect(await screen.findByText(fakeErrorText)).toBeInTheDocument();
        });

        it("Shows error dialog when share completion fails (generic error)", async () => {
            const fakeErrorText = "Fake plain error in test";
            mockClient.bindThreePid.mockRejectedValue(new Error(fakeErrorText));
            fireEvent.click(screen.getByText("Complete"));

            // Expect error dialog/modal to be shown. We have to wait for the UI to transition.
            // Check the title
            expect(await screen.findByText("Unable to verify email address.")).toBeInTheDocument();
            // Check the description
            expect(await screen.findByText(fakeErrorText)).toBeInTheDocument();
        });
    });
});

describe("<EmailAddresses />", () => {
    it("should render a loader while loading", async () => {
        const { container } = render(<EmailAddresses emails={[emailThreepidFixture]} isLoading={true} />);

        expect(container).toMatchSnapshot();
    });

    it("should render email addresses", async () => {
        const { container } = render(<EmailAddresses emails={[emailThreepidFixture]} isLoading={false} />);

        expect(container).toMatchSnapshot();
    });

    it("should handle no email addresses", async () => {
        const { container } = render(<EmailAddresses emails={[]} isLoading={false} />);

        expect(container).toMatchSnapshot();
    });
});
