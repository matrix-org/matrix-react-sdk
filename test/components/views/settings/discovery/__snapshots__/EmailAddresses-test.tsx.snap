// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailAddresses /> should handle no email addresses 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
      >
        Email addresses
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_description"
    >
      <div
        class="mx_SettingsSubsection_text"
      >
        Discovery options will appear once you have added an email above.
      </div>
    </div>
  </div>
</div>
`;

exports[`<EmailAddresses /> should render a loader while loading 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
      >
        Email addresses
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_content mx_SettingsSubsection_contentStretch"
    >
      <div
        class="mx_InlineSpinner"
      >
        <div
          aria-label="Loadingâ€¦"
          class="mx_InlineSpinner_icon mx_Spinner_icon"
          style="width: 16px; height: 16px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EmailAddresses /> should render email addresses 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
      >
        Email addresses
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_content mx_SettingsSubsection_contentStretch"
    >
      <div
        class="mx_GeneralUserSettingsTab_section--discovery_existing"
      >
        <span
          class="mx_GeneralUserSettingsTab_section--discovery_existing_address"
        >
          foo@bar.com
        </span>
        <div
          class="mx_AccessibleButton mx_GeneralUserSettingsTab_section--discovery_existing_button mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_sm"
          role="button"
          tabindex="0"
        >
          Share
        </div>
      </div>
    </div>
  </div>
</div>
`;
